From 5fa4e34f069742d32817522e5c668bfe546f8462 Mon Sep 17 00:00:00 2001
From: Tamizh Chelvam Raja <quic_tamizhr@quicinc.com>
Date: Fri, 4 Oct 2024 12:31:52 +0530
Subject: [PATCH] QSDK: wifi: ath12k: invalidate msdu before dma map in rx replenish

Do invalidate skb in rx replenish path for the buffer which is not allocated
from skb recylcer. This will avoid incorrect cache flush operation of the buffer
which may lead to msdu data corruption in long run scneario.

Signed-off-by: Tamizh Chelvam Raja <quic_tamizhr@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_rx.c | 1 +
 1 file changed, 1 insertion(+)

--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -305,6 +305,9 @@ void ath12k_dp_rx_bufs_replenish(struct
 		if (unlikely(!skb))
 			break;
 
+		if (unlikely(!skb->fast_recycled))
+			dmac_inv_range_no_dsb(skb->data, skb->data + (skb->len + skb_tailroom(skb)));
+
 		paddr = dma_map_single_attrs(ab->dev, skb->data,
 					     skb->len + skb_tailroom(skb),
 					     DMA_FROM_DEVICE, DMA_ATTR_SKIP_CPU_SYNC);
@@ -318,6 +321,7 @@ void ath12k_dp_rx_bufs_replenish(struct
 		num_remain--;
 	}
 
+	dsb(st);
 	spin_lock_bh(&srng->lock);
 	ath12k_hal_srng_access_begin(ab, srng);
 	while ((skb = __skb_dequeue(&skb_list))) {
